Add activation script to check for system-libs that override openssl install

h-vetinari opened the pull request:
<!--
Thank you for pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

This comes out of discussions with @isuruf and @mingwandroid in https://github.com/conda-forge/cryptography-feedstock/pull/37 and https://github.com/conda-forge/python-feedstock/pull/345.

The short story is that the cryptography test suite luckily picked up that the presence of outdated system libs in `C:\Windows\System32` caused a wrong openssl-version to be linked under the hood.

While setting `CONDA_DLL_SEARCH_MODIFICATION_ENABLE` allows to work around this, it was deemed "too big of a hammer" to be generally enabled. However, in order not to leave affected users (with such outdated libs on the system path) unknowingly vulnerable, the suggestion was to check this in an activation script for `openssl`.

The script works as follows:
* check which `libssl-1_1-x64.dll` is the first on PATH
* check which `libcrypto-1_1-x64.dll` is the first on PATH
* if any of the respective paths contains "System32":
    * if `CONDA_DLL_SEARCH_MODIFICATION_ENABLE` is not set: set it and try again
    * if it had been set already: display a warning to the user

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

h-vetinari commented:
@conda-forge-admin, please rerender

h-vetinari commented:
> It doesn't matter which DLL is the first on PATH. Windows loader looks at `C:\Windows\System32` first and then the paths on `PATH`. So, we should just look at whether the file `C:\Windows\System32\libcrypto-1_1-x64.dll` exists or not as I mentioned at conda-forge/cryptography-feedstock#37 (comment)

I wasn't aware of this intricacy of the windows loader. It unfortunately messes up my idea to set `CONDA_DLL_SEARCH_MODIFICATION_ENABLE` only if necessary (and test the result), because I haven't found a way to determine which DLL would be loaded first in that case (completely aside from that envvar, not even [`ctypes.util.find_library`](https://docs.python.org/3/library/ctypes.html#finding-shared-libraries) yields the correct result when `C:\Windows\System32` is involved.

I mean, running python and checking the out put of
```
from cryptography.hazmat.backends.openssl import backend
backend.openssl_version_text()
```
would work in principle (but not for this feedstock). Does someone have an idea on how I could check which library the windows loader would pick up (say, from within python).

h-vetinari commented:
Travis is just a timeout, but I don't have the rights to restart it.

h-vetinari commented:
@conda-forge-admin, please rerender

isuruf commented on recipe/activate.bat:
```diff
@@ -0,0 +1,26 @@
+  if "%HAS_SYS_SSL%"=="T"    ECHO WARNING: %LIBSSL_PATH%
+  if "%HAS_SYS_CRYPTO%"=="T" ECHO WARNING: %LIBCRYPTO_PATH%
+                             ECHO WARNING: These libraries will be linked before those in the conda
+                             ECHO WARNING: environment and might make your installation vulnerable!
```
Add a message about `CONDA_DLL_SEARCH_MODIFICATION_ENABLE` as well?

h-vetinari replied:
Well yes, but even if they do set it, we leave the user guessing whether it worked or not, which is far from ideal.

I would have liked to do that (set the envvar and try again) within the activation script, but I cannot figure out a way to reliably determine which dll get's picked up, short of loading `cryptography` (obviously not available in this recipe) and seeing which backend gets loaded.

h-vetinari replied:
How about doing that `CONDA_DLL_SEARCH_MODIFICATION_ENABLE` check upon activation in the `cryptography` feedstock?

For openssl it wouldn't help necessarily anyway, that is - if a package wants to load the dll *from outside of python*, then it would still pick up the one in `C:\Windows\System32`. Only within a correctly adapted python (cf. https://github.com/conda-forge/python-feedstock/pull/345) would the `CONDA_DLL_SEARCH_MODIFICATION_ENABLE` adapt the windows loader correctly, IIUC.

h-vetinari replied:
Ping @isuruf 
What do you think of my last comment?

isuruf replied:
Makes sense

mingwandroid replied:
I love these changes!

h-vetinari commented:
I had made an oversight while refactoring - `%LIBSSL_PATH%` was not defined. I now fixed and re-tested it (both cases), and also added the outstanding bash-activation scripts (tested with git-bash).

h-vetinari commented:
Thinking ahead to implementing a similar activation script for `cryptography` as [described](https://github.com/conda-forge/openssl-feedstock/pull/56#discussion_r416378391), I'm wondering if we have any control over the order of the activation scripts? Or should the `bld.bat` of `cryptography` overwrite/replace the activation script from `openssl` (or maybe use a shared name)?

mingwandroid commented:
Why don't we make a common package for this check that both openssl and anything else can depend on? I need to check that activation order is correct though yeah..

h-vetinari commented:
> Why don't we make a common package for this check that both openssl and anything else can depend on? I need to check that activation order is correct though yeah..

That's certainly a smart approach, but there are non-python environments that just use openssl (and should see a warning like in this PR).

But I'm guessing this could be solved with some additional checks in the activation script of that common dependency?

isuruf commented:
This warning should always be there because `CONDA_DLL_SEARCH_MODIFICATION_ENABLE` applies only to the python processes and any openssl binaries will not use it.

h-vetinari commented:
> This warning should always be there because `CONDA_DLL_SEARCH_MODIFICATION_ENABLE` applies only to the python processes and any openssl binaries will not use it.

This is pretty much what I've done in https://github.com/conda-forge/staged-recipes/pull/11452 :)
Still a draft, but should be enough as a basis for discussion.

isuruf commented:
Why not merge this as is?

h-vetinari commented:
> Why not merge this as is?

We can, but once we also add an activation script for cryptography, we'll be duplicating warnings, or having to muck around with order/dependencies of those scripts. It wasn't my idea, but I tend to agree with @mingwandroid that the cleanest solution is common dependency that covers all cases.

isuruf commented:
Why do we need another warning for cryptography?

h-vetinari commented:
> Why do we need another warning for cryptography?

Because only there can we try to activate `CONDA_DLL_SEARCH_MODIFICATION_ENABLE` and _**test**_ whether that was successful. It had certainly seemed like you had [agreed](https://github.com/conda-forge/openssl-feedstock/pull/56#discussion_r418705896) with that approach a few hours ago.

isuruf commented:
Yes, but the warning for openssl should be there regardless. So, the two scripts does two different things. `cryptography` script can use an env variable from `openssl` script to warn.

h-vetinari commented:
> Yes, but the warning for openssl should be there regardless. So, the two scripts does two different things. cryptography script can use an env variable from openssl script to warn.

That's an option, true. It's just not very user friendly because they'll be getting the same or similar warnings twice, because we (or at least I) don't know in which order the activation scripts will be triggered.

That might mean (assuming cryptography comes first and succeeds with `CONDA_DLL_SEARCH_MODIFICATION_ENABLE`):
```
C:\Users\Someone>conda activate my_env
# cryptography.activate
Warning: Found (potentially) outdated libssl/libcrypto libraries on system;
Warning: Attempting to re-activate with CONDA_DLL_SEARCH_MODIFICATION_ENABLE

Warning: Your system contains (potentially) outdated libraries under:
Warning: C:\Windows\System32\libssl-1_1-x64.dll
Warning: C:\Windows\System32\libcrypto-1_1-x64.dll
Warning: Within this environment, the python-runtime will correctly load
Warning: the openssl version of the environment, but be aware that anything
Warning: *outside* of python that is trying to load libssl/libcrypto will
Warning: load the DLLs above, which might make that application vulnerable!

# openssl.activate
WARNING: Your system contains (potentially) outdated libraries under:
WARNING: C:\Windows\System32\libssl-1_1-x64.dll
WARNING: C:\Windows\System32\libcrypto-1_1-x64.dll
WARNING: These libraries will be used before those in the conda
WARNING: environment and might make your installation vulnerable!
(my_env) > C:\Users\Someone>
```

The third warning is IMO unnecessarily LOUD, and can be avoided with the approach in https://github.com/conda-forge/staged-recipes/pull/11452. The openssl-only block would only get shown if no cryptography is installed.

isuruf commented:
> because we (or at least I) don't know in which order the activation scripts will be triggered.
I think the order is dependency order and if there's a cycle, alphabetically by the name of the script.

h-vetinari commented:
> I think the order is dependency order and if there's a cycle, alphabetically by the name of the script.

Assuming we get the order deterministically, we still need to either block or overwrite one or the other, depending on which libraries are loaded, if we want to avoid duplicated warnings. This sounds like much more pain than it's worth (especially if there's an easy alternative).

h-vetinari commented:
@isuruf 
In one aspect you have a point - to avoid cycles, https://github.com/conda-forge/staged-recipes/pull/11452 does not depend on `openssl` / `cryptography`, but still *checks*  for their presence.

Arguably then, this does not need a new common dependency, but the activation script here could just as well probe for the presence of `cryptography`, and react accordingly.

hmaarrfk commented:
Do we still need this?
